### PR TITLE
Xpetra: HIP support

### PR DESCRIPTION
**Pull request by lucbv**
@trilinos/xpetra 

## Motivation
Enabling HIP support in Xpetra to allow propagation of HIP backend to Zoltan2 and MueLu

## Related Issues

* Closes 
* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to #9322 
* Part of 
* Composed of 

## Stakeholder Feedback
ExaWind, SPARC and EMPIRE care about this feature as they are targeting AMD architectures 

## Testing
All testing was performed on Spock (OLCF pre-Frontier AMD testbed)

### CODE REVIEW

**Comment by lucbv** [comment]
Tests running on Spock:
```
ctest -R
Test project /gpfs/alpine/scratch/lberge/csc465/builds/trilinos_lucbv/packages/xpetra
      Start  1: Xpetra_Laplace1D-Xpetra_MPI_4
 1/16 Test  #1: Xpetra_Laplace1D-Xpetra_MPI_4 ......................   Passed    0.92 sec
      Start  2: Xpetra_Laplace1D-Tpetra_MPI_4
 2/16 Test  #2: Xpetra_Laplace1D-Tpetra_MPI_4 ......................   Passed    0.83 sec
      Start  3: Xpetra_Map_UnitTests_MPI_4
 3/16 Test  #3: Xpetra_Map_UnitTests_MPI_4 .........................   Passed    0.82 sec
      Start  4: Xpetra_Map_UnitTests2_MPI_4
 4/16 Test  #4: Xpetra_Map_UnitTests2_MPI_4 ........................   Passed    0.80 sec
      Start  5: Xpetra_MapFactory_UnitTests_MPI_4
 5/16 Test  #5: Xpetra_MapFactory_UnitTests_MPI_4 ..................   Passed    0.81 sec
      Start  6: Xpetra_StridedMap_UnitTests_MPI_4
 6/16 Test  #6: Xpetra_StridedMap_UnitTests_MPI_4 ..................   Passed    0.83 sec
      Start  7: Xpetra_StridedMapFactory_UnitTests_MPI_4
 7/16 Test  #7: Xpetra_StridedMapFactory_UnitTests_MPI_4 ...........   Passed    0.82 sec
      Start  8: Xpetra_MapExtractorFactory_UnitTests_MPI_4
 8/16 Test  #8: Xpetra_MapExtractorFactory_UnitTests_MPI_4 .........   Passed    0.84 sec
      Start  9: Xpetra_MultiVector_UnitTests_MPI_4
 9/16 Test  #9: Xpetra_MultiVector_UnitTests_MPI_4 .................   Passed    0.94 sec
      Start 10: Xpetra_MultiVector_UnitTestsXpetraSpecific_MPI_4
10/16 Test #10: Xpetra_MultiVector_UnitTestsXpetraSpecific_MPI_4 ...   Passed    0.82 sec
      Start 11: Xpetra_CrsMatrix_UnitTests_MPI_4
11/16 Test #11: Xpetra_CrsMatrix_UnitTests_MPI_4 ...................   Passed    0.88 sec
      Start 12: Xpetra_CrsMatrixUtils_UnitTests_MPI_4
12/16 Test #12: Xpetra_CrsMatrixUtils_UnitTests_MPI_4 ..............   Passed    0.82 sec
      Start 13: Xpetra_Matrix_UnitTests_MPI_4
13/16 Test #13: Xpetra_Matrix_UnitTests_MPI_4 ......................   Passed    0.86 sec
      Start 14: Xpetra_MatrixMatrix_UnitTests_MPI_4
14/16 Test #14: Xpetra_MatrixMatrix_UnitTests_MPI_4 ................   Passed    0.98 sec
      Start 15: Xpetra_MatrixUtils_UnitTests_MPI_4
15/16 Test #15: Xpetra_MatrixUtils_UnitTests_MPI_4 .................   Passed    0.86 sec
      Start 16: Xpetra_BlockedMultiVector_UnitTests_MPI_4
16/16 Test #16: Xpetra_BlockedMultiVector_UnitTests_MPI_4 ..........   Passed    1.04 sec

100% tests passed, 0 tests failed out of 16

Label Time Summary:
Xpetra    =  55.44 sec*proc (16 tests)

Total Test time (real) =  13.87 sec
```

File: packages/xpetra/src/Vector/Xpetra_EpetraVector.cpp [hunk 1]
```diff
@@ -128,8 +134,8 @@ template class EpetraVectorT<long long, Kokkos::Compat::KokkosOpenMPWrapperNode
 template Epetra_Vector & toEpetra<long long,Kokkos::Compat::KokkosOpenMPWrapperNode>(Vector<double, int, long long, Kokkos::Compat::KokkosOpenMPWrapperNode> &);
 template const Epetra_Vector & toEpetra<long long,Kokkos::Compat::KokkosOpenMPWrapperNode>(const Vector<double, int, long long, Kokkos::Compat::KokkosOpenMPWrapperNode> &);
 #endif
-#ifdef HAVE_TPETRA_INST_CUDA
-typedef Kokkos::Compat::KokkosCudaWrapperNode default_node_type;
+#ifdef HAVE_TPETRA_INST_HIP
```
Comment (jhux2) [inline packages/xpetra/src/Vector/Xpetra_EpetraVector.cpp:137]: Why was the `HAVE_TPETRA_INST_CUDA` clause removed?

File: packages/xpetra/src/Import/Xpetra_EpetraImport.cpp [hunk 2]
```diff
@@ -114,8 +119,8 @@ template RCP<const Import<int,long long,Kokkos::Compat::KokkosThreadsWrapperNode
 template class EpetraImportT<long long, Kokkos::Compat::KokkosOpenMPWrapperNode >;
 template RCP<const Import<int, long long, Kokkos::Compat::KokkosOpenMPWrapperNode> > toXpetra<long long,Kokkos::Compat::KokkosOpenMPWrapperNode>(const Epetra_Import *);
 #endif
-#ifdef HAVE_TPETRA_INST_CUDA
-typedef Kokkos::Compat::KokkosCudaWrapperNode default_node_type;
+#ifdef HAVE_TPETRA_INST_HIP
```
Comment (jhux2) [inline packages/xpetra/src/Import/Xpetra_EpetraImport.cpp:122]: Why was the HAVE_TPETRA_INST_CUDA clause removed?

File: packages/xpetra/src/MultiVector/Xpetra_EpetraIntMultiVector.cpp [hunk 3]
```diff
@@ -127,8 +133,8 @@ template class EpetraIntMultiVectorT<long long, Kokkos::Compat::KokkosOpenMPWrap
 template Epetra_IntMultiVector & toEpetra<long long,Kokkos::Compat::KokkosOpenMPWrapperNode >(MultiVector<int, int, long long, Kokkos::Compat::KokkosOpenMPWrapperNode> &);
 template const Epetra_IntMultiVector & toEpetra<long long, Kokkos::Compat::KokkosOpenMPWrapperNode >(const MultiVector<int, int, long long, Kokkos::Compat::KokkosOpenMPWrapperNode> &);
 #endif
-#ifdef HAVE_TPETRA_INST_CUDA
-typedef Kokkos::Compat::KokkosCudaWrapperNode default_node_type;
+#ifdef HAVE_TPETRA_INST_HIP
```
Comment (jhux2) [inline packages/xpetra/src/MultiVector/Xpetra_EpetraIntMultiVector.cpp:136]: Why was the HAVE_TPETRA_INST_CUDA clause removed?

**Review by jhux2 — CHANGES_REQUESTED** [review]
Huh, that's it?  I suspect the lack of unit tests in Xpetra is hiding things, and that MueLu tests will reveal issues.

File: packages/xpetra/src/Vector/Xpetra_EpetraVector.cpp (hunk 1 above)
Comment (lucbv) [reply packages/xpetra/src/Vector/Xpetra_EpetraVector.cpp:143]: Fixed

File: packages/xpetra/src/Import/Xpetra_EpetraImport.cpp (hunk 2 above)
Comment (lucbv) [reply packages/xpetra/src/Import/Xpetra_EpetraImport.cpp:127]: Fixed

File: packages/xpetra/src/MultiVector/Xpetra_EpetraIntMultiVector.cpp (hunk 3 above)
Comment (lucbv) [reply packages/xpetra/src/MultiVector/Xpetra_EpetraIntMultiVector.cpp:142]: Also fixed

**Comment by lucbv** [comment]
@jhux2 yeah I am also pretty afraid because of the lack of testing.
I did grep for "cuda" and tried to modify things around all the references to it.
At least the copy/past/modify mistakes I made are taken care of. The faster we get to MueLu the better : )

**Review by jhux2 — APPROVED** [review]
Thanks, @lucbv.  Looks good to me.

**Pull request merged**